features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/OpennmsKafkaProducer.java

```diff
@@ -408,6 +409,15 @@ private void processKafkaSendQueue() {
                     producer.send(producerRecord, (recordMetadata, e) -> {
                         if (e != null) {
                             LOG.warn("Failed to send record to producer: {}.", producerRecord, e);
+                            if (e instanceof TimeoutException) {
+                                // If Kafka is Offline, buffer the record again for events.
+                                // This is best effort to keep the order although in-flight elements may still miss the order.
+                                if (kafkaRecord.producerRecord != null &&
+                                        kafkaRecord.producerRecord.topic() != null &&
+                                        this.eventTopic.equalsIgnoreCase(kafkaRecord.producerRecord.topic())) {
+                                    kafkaSendQueue.offerFirst(kafkaRecord);
```

Comment:
This is a blocking call, this may be a non-issue but it might not be desirable to block inside the callback depending on how the caller is implemented.

Reply:
I should clarify, its blocking while acquiring the lock to the queue which looks like it isn't fair. I suspect it is probably fine though, maybe in practice that will always be quick.

Reply:
The callback is just for ack. Don't think caller will have any dependency on it. 
From javadoc `offer` : Inserts the specified element into this queue if it is possible to do so immediately without violating capacity restrictions.
I think it should be similar for `offerFIrst`. It shouldn't wait for any resources to be available.

Reply:
My point was just that operations on the blockingqueue have to take the lock so they are blocking by nature. Have to lock the queue to check if there is capacity and then to insert. Its probably not an issue here I've just seen it be a problem before calling blocking operations in a callback depending on what thread it ends up executing on.

The queue is also not fair so it could block for a while potentially. Just something to keep in mind.